### ISSUE TITLE
[libc++] `std::counting_semaphore<-1>` should be rejected

### ISSUE THREAD

**Issue by frederick-vs-ja** [llvm/llvm-project#63131]
According to [[thread.sema.cnt]/2](https://eel.is/c++draft/thread.sema.cnt#2), the following program is ill-formed and thus shouldn't compile.

```C++
#include <semaphore>

int main()
{
    std::counting_semaphore<-1> sm [[maybe_unused]] {-1}; // negative least_max_value
}
```

But currently libc++ accepts it ([Godbolt link](https://godbolt.org/z/vMcoK5sss)).

A similar issue has been reported for MSVC STL (microsoft/STL#3746).

**Comment by schwarzschild-radius** [comment]
@frederick-vs-ja I think the issue has been fixed by [commit 42a024f](https://github.com/llvm/llvm-project/commit/42a024fad9fca3fbe37afd55681e84dc60dd1abe) and I can see that the following assertion emitted by clang

```
/opt/compiler-explorer/clang-trunk-20230725/bin/../include/c++/v1/semaphore:147:19: error: static assertion failed due to requirement '-1L >= 0': The least maximum value must be a positive number
  147 |     static_assert(__least_max_value >= 0, "The least maximum value must be a positive number");
      |                   ^~~~~~~~~~~~~~~~~~~~~~
<source>:5:33: note: in instantiation of template class 'std::counting_semaphore<-1>' requested here
    5 |     std::counting_semaphore<-1> sm [[maybe_unused]] {-1};
      |                                 ^
1 error generated.
```

Can this issue be closed?

**Comment by frederick-vs-ja** [comment]
> @frederick-vs-ja / @philnik777 I think the issue has been fixed by [commit 42a024f](https://github.com/llvm/llvm-project/commit/42a024fad9fca3fbe37afd55681e84dc60dd1abe) and I can see that the following assertion emitted by clang
> 
> ```
> /opt/compiler-explorer/clang-trunk-20230725/bin/../include/c++/v1/semaphore:147:19: error: static assertion failed due to requirement '-1L >= 0': The least maximum value must be a positive number
>   147 |     static_assert(__least_max_value >= 0, "The least maximum value must be a positive number");
>       |                   ^~~~~~~~~~~~~~~~~~~~~~
> <source>:5:33: note: in instantiation of template class 'std::counting_semaphore<-1>' requested here
>     5 |     std::counting_semaphore<-1> sm [[maybe_unused]] {-1};
>       |                                 ^
> 1 error generated.
> ```
> 
> Can this issue be closed?

Yes. Let me close this. (Although the diagnostic message is a bit inaccurate...)
